### PR TITLE
487 comments in-between if else blocks

### PR DESCRIPTION
**Pull request by yamafaktory**

### CODE REVIEW

**Comment by yamafaktory** [comment]
This one should solve #487 🤞.

File: src/comments.js [hunk 1]
```diff
@@ -153,8 +154,12 @@ function attach(comments, ast, text) {
       // We also need to check if it's the first line of the file.
 
       if (followingNode) {
-        // Always a leading comment.
-        addLeadingComment(followingNode, comment);
+        // There are some exceptions like in-between comments.
+        if (enclosingNode && enclosingNode.type === 'IfStatement') {
```
Comment (jlongster) [inline src/comments.js]: Hm, this comment code is already very complicated so I'd love to find a way to that doesn't special-case `if`s in here. This code is going to become unmaintainable if we allow special-cases; even without them it's already very hard to understand.

File: src/comments.js [hunk 2]
```diff
@@ -332,7 +341,15 @@ function printTrailingComment(commentPath, print, options, parentNode) {
     // trailing comment for `2`. We can simulate the above by checking
     // if this a comment on its own line; normal trailing comments are
     // always at the end of another expression.
-    return concat([hardline, contents]);
+    if (isException) {
```
Comment (jlongster) [inline src/comments.js]: Same concern here, the `print{Leading,Trailing}Comment` functions are written very carefully and I don't want to make them any more complicated. Let me think about this a bit.

Comment (yamafaktory) [inline src/comments.js]: Thanks for the review, I know those in-between comments are introducing a somehow different logic and it was hard not to make this over-complex. Hope you will find a better (and simpler) implementation 👍.

**Comment by jlongster** [comment]
Thanks for trying this out! Let's discuss over in the issue, I just commented here: https://github.com/jlongster/prettier/issues/487#issuecomment-276090434

**Comment by vjeux** [comment]
As mentioned on #487, let's move the comments that are in the middle of an `if else` to the beginning of the else block. @yamafaktory, do you think you could implement this behavior? Thanks!

**Comment by vjeux** [comment]
```js
if (1) false
// comments
else true
```

already outputs

```js
if (1)
  false;
else
  // vjeux
  true;
```

so the only case that needs to be handled is moving the comment inside of the block instead of leading it.

**Comment by yamafaktory** [comment]
@vjeux Yes, I can do it 👍.

**Comment by vjeux** [comment]
Adding more test cases to support.

```js
if (1)
// comment
{
  false
}
// comment
else if (1) {
  true
}
// comment
else {
	
}
```

**Comment by vjeux** [comment]
I have something kind of working

```js
diff --git a/src/comments.js b/src/comments.js
index e05c7f5..e5145bb 100644
--- a/src/comments.js
+++ b/src/comments.js
@@ -152,7 +152,15 @@ function attach(comments, ast, text) {
       // If a comment exists on its own line, prefer a leading comment.
       // We also need to check if it's the first line of the file.

-      if (followingNode) {
+      if (
+        enclosingNode.type === "IfStatement" &&
+        followingNode.type === "BlockStatement") {
+        if (followingNode.body.length === 0) {
+          addDanglingComment(followingNode, comment);
+        } else {
+          addLeadingComment(followingNode.body[0], comment);
+        }
+      } else if (followingNode) {
         // Always a leading comment.
         addLeadingComment(followingNode, comment);
       } else if (precedingNode) {
```

```js
if (1) {
  // comment
  false;
} else // comment
if (1) {
  true;
} else {
  // comment
}
```

but it's still missing the else if case. I think it's close though.

**Comment by vjeux** [comment]
Fixed in #544, thanks a lot for the inspiration!

**Pull request closed without merging**